### DISCUSSION TITLE
Does there exist a way to get all "pure descendants" of a node?

### DISCUSSION THREAD

**Discussion by ghost** [networkx/networkx#7151] [Q&A]
The name "pure descendant" is something I made up for lack of a better name, couldn't find an official name or description for this concept.

Definition of "pure descendant": A node N is a pure descendant of reference node R if: ancestors(N) - ancestors(R) - descendants(R) is an empty set. To put it another way, a node N is a pure descendant of R if all of node N's direct parents are either a pure descendant of or an ancestor of R. To clarify, it is not sufficient to say that all of N's direct parents are a descendant of R, since an ancestor of R could be a direct parent of N, and the condition still be satisfied.

How to do it (one possibility): get set of R's ancestors, create empty set of pure descendants. Traverse the descendants of R in topologically sorted order, at each node N, if all of N's parents are in the pure descendants set or in R's ancestor set, add the node to the pure descendants set.

If this functionality doesn't already exist in the library, would the team consider adding this under dag algorithms?

**Comment by rossbar** [comment]
> A node N is a pure descendant of reference node R if: ancestors(N) - ancestors(R) - descendants(R) is an empty set.

It sounds like this is the answer to your question!

```python
nx.ancestors(G, N) - nx.ancestors(G, R) - nx.descendants(G, R) == {}
```

**Reply by ghost** [reply]
The problem with this approach is that it will only tell you for a single target node if it is a pure descendant of a reference node. To extrapolate this to get all pure descendants using this method you'd need to do this calc for every descendant of R. The anc/desc of R would remain constant, but you'd need to call ancestors(R) for every node, which wouldn't be very efficient. It also get's trickier when you want to get pure_descendants of multiple source nodes together, since `pure_descendants(N1) + pure_descendants(N2) != pure_descendants(G, N1 + N2)`

The approach I stated in the last paragraph would be a more preferable way I think. Adding a quick implementation below. This approach takes multiple source nodes and finds the pure descendants with respect to all source nodes provided.

I would propose adding this to the library of dag algorithms, as I think it's a novel algorithm with real world applications and not simple enough that it can be derived easily on the fly using existing networkx functionality.

```
def pure_descendants(G, sources):
    all_source_ancestors = set()
    all_source_descendants = set()
    for source in sources:
        all_source_ancestors |= nx.ancestors(G, source)
        all_source_descendants |= nx.descendants(G, source)

    pure_descendants = set()
    
    topo_order = nx.topological_sort(G)
    for target in topo_order:
        if target not in all_source_descendants:
            continue
        target_parents = {edge[0] for edge in G.in_edges(target)}
        if not (target_parents - (pure_descendants | all_source_ancestors | set(sources))):
            pure_descendants.add(target)
    return pure_descendants
```

**Reply by rossbar** [reply]
> The anc/desc of R would remain constant, but you'd need to call ancestors(R) for every node, which wouldn't be very efficient. 

I think this is an implementation detail - there's always a tradeoff between memory and computation. If I were tasked with doing this computation over the whole graph, I'd probably start by pre-computing the ancestors and descendants for each node. Maybe something like:

```python
>>> G = nx.DiGraph([(0, 1), (1, 2), (2, 3), (3, 4)])
>>> sources = list(G)  # or a subset of nodes
>>> ancestors = {n: nx.ancestors(G, n) for n in G}
>>> descendants = {n: nx.descendants(G, n) for n in G}
>>> pure_descendants = defaultdict(set)
>>> for R in sources:
...     for N in set(ancestors) - {R}:  # Excluding the reference node itself
...         if ancestors[N] - ancestors[R] - descendants[R] == set():
...             pure_descendants[R] |= {N}
```

where the for-loop is just the computation you've suggested as the definition applied over the specified nodes.

Re: adding a function for this to the DAG package - I'm certainly no expert but I think some things that would make a strong case are: 1) a set of clear, concise tests cases that illustrate desired behavior for common patterns in DAGs, and 2) any references to this problem/algorithm in literature.

**Reply by ghost** [reply]
I think a more clear way to think about this algorithm is a partial topological sort. This would use the standard top down bfs style algorithm, but instead of pre-seeding the queue with all leaf nodes in the graph, you pre-seed it with your sources. Additionally, you would decrement the initial indegree map for all of the direct children of all of the sources ancestors. This will result in a partial topological sort, including only the desired pure descendants of your sources.

Here is an example of how this could be done (slightly reworking the topological_generations algorithm). I think this would be the ideal way to do this.

```
def pure_descendants(G, sources):
    """
    """
    if not G.is_directed():
        raise nx.NetworkXError("Topological sort not defined on undirected graphs.")

    sources_ancestors = set()
    for source in sources:
        sources_ancestors |= nx.ancestors(G, source)
    
    indegree_map = {v: d for v, d in G.in_degree() if d > 0}

    for node in sources_ancestors:
        for child in G.neighbors(node):
            indegree_map[child] -= 1

    zero_indegree = sources

    while zero_indegree:
        this_generation = zero_indegree
        zero_indegree = []
        for node in this_generation:
            if node not in G:
                raise RuntimeError("Graph changed during iteration")
            for child in G.neighbors(node):
                try:
                    indegree_map[child] -= 1
                except KeyError as err:
                    raise RuntimeError("Graph changed during iteration") from err
                if indegree_map[child] == 0:
                    zero_indegree.append(child)
                    del indegree_map[child]
        if this_generation != sources:
            yield from this_generation
```

Note, this could also be done using graphlib.TopologicalSorter. First you get all ancestors of your sources. Mark them all done in the sorter object. Then from there, mark your sources done and while sorter.get_ready() returns anything you add them to the result, mark them done and repeat until sorter.get_ready() doesn't return anything. 

For your suggestions for making a case for adding this algorithm for (1) please find below some illustrations for examples of what this algorithm would produce (by no means a full set of test cases yet). Btw I believe the example solution you provided, doesn't quite provide the expected result, which is clearer now that the expectations are illustrated. For (2) I can't quite find any literature regarding this concept, but if I can I'd like to point to an example of the previous addition of topological_generations (https://github.com/networkx/networkx/pull/4757). When we added this I did not identify any literature regarding this concept either, is this a requirement we can forgo in considering this algorithm?

<img width="300" alt="Screen Shot 2023-12-13 at 7 45 04 PM" src="https://github.com/networkx/networkx/assets/13501084/075fbc18-1ea9-4037-a68a-32b6dfad51ee">
<img width="277" alt="Screen Shot 2023-12-13 at 7 46 39 PM" src="https://github.com/networkx/networkx/assets/13501084/8c57d7e8-1b8a-4c7a-bf1c-d90b1ca3c050">
<img width="288" alt="Screen Shot 2023-12-13 at 7 47 06 PM" src="https://github.com/networkx/networkx/assets/13501084/2b993887-dc0a-4442-8315-491c73fb4447">
<img width="310" alt="Screen Shot 2023-12-13 at 8 44 28 PM" src="https://github.com/networkx/networkx/assets/13501084/3050f49e-add2-453a-8e94-374499dcab67">

**Reply by ghost** [reply]
actually this could even just be a slight generalization of the topological_generations algorithm itself. The only real difference being if sources provided then you do the extra if condition before the while loop (and don't raise at the end if indgree_map not empty)

```
def topological_generations(G, sources=None):
    if not G.is_directed():
        raise nx.NetworkXError("Topological sort not defined on undirected graphs.")

    multigraph = G.is_multigraph()
    indegree_map = {v: d for v, d in G.in_degree() if d > 0}

    if sources is None:
        zero_indegree = [v for v, d in G.in_degree() if d == 0]
    else:
        # just adding the logic in this else condition  <-----------------------------------
        sources_ancestors = set()
        for source in sources:
            sources_ancestors |= nx.ancestors(G, source)
        for node in sources_ancestors:
            for child in G.neighbors(node):
                indegree_map[child] -= len(G[node][child]) if multigraph else 1
        zero_indegree = sources

    while zero_indegree:
        this_generation = zero_indegree
        zero_indegree = []
        for node in this_generation:
            if node not in G:
                raise RuntimeError("Graph changed during iteration")
            for child in G.neighbors(node):
                try:
                    indegree_map[child] -= len(G[node][child]) if multigraph else 1
                except KeyError as err:
                    raise RuntimeError("Graph changed during iteration") from err
                if indegree_map[child] == 0:
                    zero_indegree.append(child)
                    del indegree_map[child]
        yield this_generation

    if indegree_map and sources is None:  # just adding this sources check <--------------------------------
        raise nx.NetworkXUnfeasible(
            "Graph contains a cycle or graph changed during iteration"
        )


def topological_sort(G, sources=None):  # add optional sources here, pass to topological_gens below <-----------------
    for generation in nx.topological_generations(G, sources=sources):
        yield from generation

def pure_descendants(G, sources):  # optional additional function
    return set(topological_sort(G, sources))
```

**Comment by ghost** [comment]
@rossbar @dschult - any thoughts on the above proposal (https://github.com/networkx/networkx/discussions/7151#discussioncomment-7849296)? would be very minimal modifications to existing algorithms, and require no structural changes.

**Reply by ghost** [reply]
Hi and happy new year! I actually thought about this a bit more and I think the desired behavior is even simpler than described earlier. The last image above is incorrect. Ancestors of the reference node should not be considered as a “satisfied” dependency. So the last example should actually have no pure descendants. This makes the modification to topological_generations even more trivial:

```
if sources is None:
    zero_indegree = [v for v, d in G.in_degree() if d == 0]
else:
    zero_indegree = sources
```
